cou/utils/juju_utils.py

```diff
@@ -168,27 +168,27 @@ def name(self) -> str:
         return self._name
 
     @retry(no_retry_exceptions=(ActionFailed,))
-    async def _get_action_result(self, action: Action, raise_on_failure: bool) -> Action:
-        """Get results from action.
+    async def _get_waited_action_object(self, action: Action, raise_on_failure: bool) -> Action:
+        """Get waited action object.
+
+        To access action data from the returned action object, use `action_obj.data`, which
+        contains action parameters, results, status, and metadata. Alternatively, it is possible
+        to access action results directly with `action_obj.results`.
 
         :param action: Action object
         :type: Action
         :param raise_on_failure: Whether to raise ActionFailed exception on failure, defaults
                                  to False
         :type raise_on_failure: bool
-        :return: action results
+        :return: the awaited action object
         :rtype: Action
         :raises ActionFailed: When the application status is in error (it's not 'completed').
         """
-        result = await action.wait()
-        if raise_on_failure:
-            model = await self._get_model()
-            status = await model.get_action_status(uuid_or_prefix=action.entity_id)
-            if status != "completed":
-                output = await model.get_action_output(action.entity_id)
-                raise ActionFailed(action, output=output)
+        action_obj = await action.wait()
+        if raise_on_failure and action_obj.status != "completed":
+            raise ActionFailed(action, output=action_obj.data)
```

Comment:
Previous approach was chosen by OpenStak team in zaza, I'm not sure why. @gabrielcocenza do you know if it's important?

Reply:
I just checked zaza, they are doing the [same type of comparison](https://github.com/openstack-charmers/zaza/blob/master/zaza/model.py#L1099) (I'm surprised that they are even using the same variable name as mine :laughing: ). The only difference is how they are getting the output, which I don't know if there's any additional benefits getting from model instead of the action itself.

Reply:
I think that this is fine, I have no idea why it was like this in the first place, and I like your version.

Reply:
I have no idea....hahahaha I also enjoyed the approach :+1: